tensorflow_datasets/core/features/shaped_feature.py

```diff
@@ -0,0 +1,190 @@
+"""Provides `StaticShapedTensor` and `DynamicShapedTensor`.
+
+These are designed for composition with other `FeatureConnector`s whose internal
+representation is unable to represent shape, e.g. run length encoded features.
+"""
+
+import numpy as np
+import tensorflow as tf
+import tensorflow_datasets as tfds
+from tensorflow_datasets.core.features import feature
+
+
+def _assert_is_feature(connector, name):
+  if not isinstance(connector, feature.FeatureConnector):
+      raise ValueError(
+          "%s must be a `FeatureConnector` but got %s"
+          % (name, connector))
+
+
+def _num_unknown_dims(shape):
+  return shape.count(None)
+
+
+class DynamicShapedTensor(feature.FeaturesDict):
```

Comment:
Are these 2 new connectors maybe better to add to the `Tensor` `FeatureConnector`. There's a [todo](https://github.com/tensorflow/datasets/blob/d2123d21ec68607a0de830ab3afcde36162190f0/tensorflow_datasets/core/features/feature.py#L566) to support dynamic shapes.

If the `shape` passed to the `Tensor` constructor was not fully known, then you could write out to disk the feature value as well as the shape and reshape on the way out in decode.

Also note that you should not subclass `FeaturesDict` here. `FeatureConnector`s that are logically just a single features can still write multiple features to disk. `get_serialized_info` would reflect the fact that on disk there are multiple fields connected to this feature.

Reply:
This is what I've attempted to do here (originally there was even a `ShapedTensor` function that directed appropriately depending on the `shape`).

I'm happy to move away from `FeaturesDict` (didn't realize similar behaviour was supported by `FeatureConnector`s), but I'm unsure how to implement this directly with `Tensor`. I could subclass `Tensor` and redict to the super method purely for the reshape, but that feels awfully convoluted and conceptually wrong. (B)RLE encodings are intrinsically 1D, and my gut is much happier composing ND features out of an encoding feature and a shape feature than duplicating classes or writing code with lots of `if else` clauses.